Regression in v0.6.6-recidiviz.3.5: Positional query parameters broken due to allow_undeclared_parameters

### Description

Version `0.6.6-recidiviz.3.5` (released Feb 9, 2025) introduced a regression that breaks positional query parameters in JDBC prepared statements. The emulator now returns an error: `"Positional query parameters cannot be provided when undeclared parameters are allowed"`.

This appears to be related to ZetaSQL's `allow_undeclared_parameters` setting being enabled, which [according to ZetaSQL documentation](https://pkg.go.dev/github.com/nutshelllabs/go-zetasql), blocks the use of positional parameters.

---

### Steps to Reproduce

1. Create a prepared statement with positional parameters:
```java
String sql = "SELECT name FROM users WHERE age > ?";
PreparedStatement pstmt = connection.prepareStatement(sql);
pstmt.setInt(1, 25);
ResultSet rs = pstmt.executeQuery();
```

2. Execute the query using the BigQuery Java SDK with `QueryJobConfiguration`:
```java
QueryJobConfiguration config = QueryJobConfiguration.newBuilder(sql)
    .setUseLegacySql(false)
    .setPositionalParameters(parameters)
    .build();
```

3. Query execution fails with error

---

### Expected Behavior

Positional parameters should work as they did in v0.6.6-recidiviz.3.4. The query should execute successfully with the provided parameter values.

---

### Actual Behavior

Query fails with:
```
INVALID_ARGUMENT: Positional query parameters cannot be provided when undeclared parameters are allowed
failed to configure query parameter types: INVALID_ARGUMENT: Positional query parameters cannot be provided when undeclared parameters are allowed
```

Full stack trace:
```
com.google.cloud.bigquery.BigQueryException: failed to configure query parameter types: INVALID_ARGUMENT: Positional query parameters cannot be provided when undeclared parameters are allowed
	at com.google.cloud.bigquery.BigQueryRetryHelper.runWithRetries(BigQueryRetryHelper.java:71)
	at com.google.cloud.bigquery.BigQueryImpl.getQueryResults(BigQueryImpl.java:2150)
	...
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
GET http://localhost:9050/bigquery/v2/projects/test-project/queries/{jobId}?maxResults=0&prettyPrint=false
{
  "code": 400,
  "errors": [
    {
      "location": "",
      "message": "failed to configure query parameter types: INVALID_ARGUMENT: Positional query parameters cannot be provided when undeclared parameters are allowed",
      "reason": "jobInternalError",
      "debugInfo": ""
    }
  ]
}
```

---

### Version Information

- **Working version:** `ghcr.io/recidiviz/bigquery-emulator:0.6.6-recidiviz.3.4`
- **Broken version:** `ghcr.io/recidiviz/bigquery-emulator:0.6.6-recidiviz.3.5` (`:latest`)
- **Change that likely introduced issue:** [Commit: "fix: allow null parameter types; loosen QueryParameterValue marshalling"](https://github.com/Recidiviz/bigquery-emulator/compare/v0.6.6-recidiviz.3.4...v0.6.6-recidiviz.3.5)

---

### Root Cause Analysis

The changes in v0.6.6-recidiviz.3.5 to fix https://github.com/goccy/bigquery-emulator/issues/137 appear to have inadvertently enabled ZetaSQL's `allow_undeclared_parameters` mode. According to [ZetaSQL documentation](https://pkg.go.dev/github.com/nutshelllabs/go-zetasql):

> "When allow_undeclared_parameters is true, no positional parameters may be provided."

This is a strict constraint in ZetaSQL/GoogleSQL that prevents using positional parameters when undeclared parameter mode is active.

---

### Impact

This regression breaks:
- All JDBC PreparedStatement queries with positional parameters (`?`)
- Standard BigQuery parameterized query patterns used by Java applications
- Integration tests for JDBC drivers (our test suite went from 108 passing to 23 failing)

---

### Workaround

Pin to the previous version until fixed:
```yaml
image: ghcr.io/recidiviz/bigquery-emulator:0.6.6-recidiviz.3.4
```

---

### Suggested Fix

The fix in v0.6.6-recidiviz.3.5 should be modified to:
1. Not enable `allow_undeclared_parameters` mode by default, OR
2. Only enable it when named parameters are used (not positional), OR
3. Provide a configuration flag to control this behavior

The goal is to maintain the Node.js parameter handling improvements from goccy#137 while keeping positional parameters functional for other clients (Java, Python, etc.).

---

### Test Case

A minimal test case demonstrating the issue is available in our integration test suite:
- Repository: https://github.com/timveil/tbc-bq-jdbc
- Test file: `src/test/java/vc/tbc/bq/jdbc/integration/ParameterizedQueryTest.java`
- Failing tests: All tests in `ParameterizedQueryTest`, `PreparedStatementAdvancedTest`, and `SessionTest` that use positional parameters

--- a/go.mod
+++ b/go.mod
@@ -118,4 +118,4 @@ require (
 	modernc.org/sqlite v1.37.0
 )
 
-replace github.com/goccy/go-zetasqlite => github.com/Recidiviz/go-zetasqlite v0.18.0-recidiviz.20
+replace github.com/goccy/go-zetasqlite => github.com/Recidiviz/go-zetasqlite v0.18.0-recidiviz.20.1

--- a/server/server_test.go
+++ b/server/server_test.go
@@ -4143,3 +4143,207 @@ func TestViewSchemaHydration(t *testing.T) {
 		}
 	}
 }
+
+// TestQueryWithPositionalParameters tests issue #69: https://github.com/Recidiviz/bigquery-emulator/issues/69
+// Verifies that positional query parameters (?) work correctly and are not broken by allow_undeclared_parameters mode.
+// The issue reports that v0.6.6-recidiviz.3.5 broke positional parameters because allow_undeclared_parameters was
+// enabled globally. According to ZetaSQL docs: "When allow_undeclared_parameters is true, no positional parameters may be provided."
+func TestQueryWithPositionalParameters(t *testing.T) {
+	const (
+		projectID = "test"
+		datasetID = "test_dataset"
+		tableID   = "test_table"
+	)
+
+	ctx := context.Background()
+
+	bqServer, err := server.New(server.TempStorage)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create test table with sample data for testing positional parameters
+	project := types.NewProject(
+		projectID,
+		types.NewDataset(
+			datasetID,
+			types.NewTable(
+				tableID,
+				[]*types.Column{
+					types.NewColumn("id", types.INTEGER),
+					types.NewColumn("name", types.STRING),
+					types.NewColumn("value", types.FLOAT),
+					types.NewColumn("active", types.BOOLEAN),
+				},
+				types.Data{
+					{"id": 1, "name": "Alice", "value": 10.5, "active": true},
+					{"id": 2, "name": "Bob", "value": 20.7, "active": false},
+					{"id": 3, "name": "Charlie", "value": 30.2, "active": true},
+					{"id": 4, "name": "David", "value": 40.9, "active": false},
+					{"id": 5, "name": "Eve", "value": 50.1, "active": true},
+				},
+			),
+		),
+	)
+	if err := bqServer.Load(server.StructSource(project)); err != nil {
+		t.Fatal(err)
+	}
+
+	testServer := bqServer.TestServer()
+	defer func() {
+		testServer.Close()
+		bqServer.Close()
+	}()
+
+	client, err := bigquery.NewClient(
+		ctx,
+		projectID,
+		option.WithEndpoint(testServer.URL),
+		option.WithoutAuthentication(),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+
+	t.Run("Multiple positional parameters", func(t *testing.T) {
+		query := client.Query(fmt.Sprintf("SELECT id, name, value FROM `%s.%s.%s` WHERE id >= ? AND id <= ? ORDER BY id", projectID, datasetID, tableID))
+		query.Parameters = []bigquery.QueryParameter{
+			{Value: 2},
+			{Value: 4},
+		}
+
+		it, err := query.Read(ctx)
+		if err != nil {
+			t.Fatalf("Query failed: %v", err)
+		}
+
+		var rows [][]bigquery.Value
+		for {
+			var row []bigquery.Value
+			if err := it.Next(&row); err != nil {
+				if err == iterator.Done {
+					break
+				}
+				t.Fatal(err)
+			}
+			rows = append(rows, row)
+		}
+
+		if len(rows) != 3 {
+			t.Fatalf("Expected 3 rows, got %d", len(rows))
+		}
+		// Verify we got ids 2, 3, 4
+		expectedIDs := []int64{2, 3, 4}
+		for i, row := range rows {
+			if row[0].(int64) != expectedIDs[i] {
+				t.Errorf("Row %d: expected id=%d, got %v", i, expectedIDs[i], row[0])
+			}
+		}
+	})
+
+	t.Run("Positional parameters with different types", func(t *testing.T) {
+		query := client.Query(fmt.Sprintf("SELECT id, name FROM `%s.%s.%s` WHERE value > ? AND active = ? ORDER BY id", projectID, datasetID, tableID))
+		query.Parameters = []bigquery.QueryParameter{
+			{Value: 25.0}, // Float
+			{Value: true}, // Boolean
+		}
+
+		it, err := query.Read(ctx)
+		if err != nil {
+			t.Fatalf("Query failed: %v", err)
+		}
+
+		var rows [][]bigquery.Value
+		for {
+			var row []bigquery.Value
+			if err := it.Next(&row); err != nil {
+				if err == iterator.Done {
+					break
+				}
+				t.Fatal(err)
+			}
+			rows = append(rows, row)
+		}
+
+		// Should get Charlie (value=30.2, active=true) and Eve (value=50.1, active=true)
+		if len(rows) != 2 {
+			t.Fatalf("Expected 2 rows, got %d", len(rows))
+		}
+		expectedNames := []string{"Charlie", "Eve"}
+		for i, row := range rows {
+			if row[1].(string) != expectedNames[i] {
+				t.Errorf("Row %d: expected name='%s', got %v", i, expectedNames[i], row[1])
+			}
+		}
+	})
+
+	t.Run("Positional parameter in LIMIT clause", func(t *testing.T) {
+		query := client.Query(fmt.Sprintf("SELECT id, name FROM `%s.%s.%s` ORDER BY id LIMIT ?", projectID, datasetID, tableID))
+		query.Parameters = []bigquery.QueryParameter{
+			{Value: 2},
+		}
+
+		it, err := query.Read(ctx)
+		if err != nil {
+			t.Fatalf("Query failed: %v", err)
+		}
+
+		var rows [][]bigquery.Value
+		for {
+			var row []bigquery.Value
+			if err := it.Next(&row); err != nil {
+				if err == iterator.Done {
+					break
+				}
+				t.Fatal(err)
+			}
+			rows = append(rows, row)
+		}
+
+		if len(rows) != 2 {
+			t.Fatalf("Expected 2 rows, got %d", len(rows))
+		}
+		// Should get first 2 rows: Alice and Bob
+		if rows[0][1].(string) != "Alice" {
+			t.Errorf("Expected first row name='Alice', got %v", rows[0][1])
+		}
+		if rows[1][1].(string) != "Bob" {
+			t.Errorf("Expected second row name='Bob', got %v", rows[1][1])
+		}
+	})
+
+	t.Run("Positional parameter with string type", func(t *testing.T) {
+		query := client.Query(fmt.Sprintf("SELECT id, name FROM `%s.%s.%s` WHERE name = ? ORDER BY id", projectID, datasetID, tableID))
+		query.Parameters = []bigquery.QueryParameter{
+			{Value: "Bob"},
+		}
+
+		it, err := query.Read(ctx)
+		if err != nil {
+			t.Fatalf("Query failed: %v", err)
+		}
+
+		var rows [][]bigquery.Value
+		for {
+			var row []bigquery.Value
+			if err := it.Next(&row); err != nil {
+				if err == iterator.Done {
+					break
+				}
+				t.Fatal(err)
+			}
+			rows = append(rows, row)
+		}
+
+		if len(rows) != 1 {
+			t.Fatalf("Expected 1 row, got %d", len(rows))
+		}
+		if rows[0][0].(int64) != 2 {
+			t.Errorf("Expected id=2, got %v", rows[0][0])
+		}
+		if rows[0][1].(string) != "Bob" {
+			t.Errorf("Expected name='Bob', got %v", rows[0][1])
+		}
+	})
+}

--- a/test/node/src/parameterizedQueries.test.ts
+++ b/test/node/src/parameterizedQueries.test.ts
@@ -518,4 +518,54 @@ describe('Parameterized Queries (Issue #58)', () => {
       expect(rows).toHaveLength(5);
     });
   });
+
+  describe('Issue #69: Positional Parameters', () => {
+    beforeAll(async () => {
+      // Create a test table for positional parameter testing
+      await helper.createTable(
+        { datasetId: 'test_dataset', tableId: 'positional_test' },
+        {
+          fields: [
+            { name: 'id', type: 'INTEGER', mode: 'REQUIRED' },
+            { name: 'name', type: 'STRING', mode: 'REQUIRED' },
+          ],
+        }
+      );
+
+      // Insert test data
+      await helper.insertRows(
+        { datasetId: 'test_dataset', tableId: 'positional_test' },
+        [
+          { id: 1, name: 'Alice' },
+          { id: 2, name: 'Bob' },
+          { id: 3, name: 'Charlie' },
+        ]
+      );
+    });
+
+    it('should handle positional parameter (?) in WHERE clause', async () => {
+      // Tests resolution of https://github.com/Recidiviz/bigquery-emulator/issues/69
+      // Verifies that positional query parameters work correctly and are not broken by allow_undeclared_parameters.
+      // According to ZetaSQL docs: "When allow_undeclared_parameters is true, no positional parameters may be provided."
+
+      const query = `
+        SELECT id, name
+        FROM \`${BQ_EMULATOR_PROJECT_ID}.test_dataset.positional_test\`
+        WHERE id = ?
+        ORDER BY id
+      `;
+
+      // Positional parameters are passed as an array without names
+      const options = {
+        query,
+        params: [2], // Positional parameter value
+      };
+
+      const [rows] = await client.query(options);
+
+      expect(rows).toHaveLength(1);
+      expect(rows[0].id).toBe(2);
+      expect(rows[0].name).toBe('Bob');
+    });
+  });
 });

--- a/test/python/emulator_test.py
+++ b/test/python/emulator_test.py
@@ -1797,3 +1797,52 @@ FROM UNNEST([
             ],
             job_config=job_config,
         )
+
+    def test_positional_query_parameters(self) -> None:
+        """Tests resolution of https://github.com/Recidiviz/bigquery-emulator/issues/69
+
+        Tests that positional query parameters (?) work correctly and are not broken
+        by allow_undeclared_parameters mode. The issue reported that v0.6.6-recidiviz.3.5
+        broke positional parameters because allow_undeclared_parameters was enabled globally.
+        According to ZetaSQL docs: "When allow_undeclared_parameters is true, no positional
+        parameters may be provided."
+        """
+        address = BigQueryAddress(dataset_id=_DATASET_1, table_id="positional_params_test")
+        self.create_mock_table(
+            address,
+            schema=[
+                bigquery.SchemaField("id", bigquery.enums.SqlTypeNames.INTEGER.value, mode="REQUIRED"),
+                bigquery.SchemaField("name", bigquery.enums.SqlTypeNames.STRING.value, mode="REQUIRED"),
+            ],
+        )
+        self.load_rows_into_table(
+            address,
+            data=[
+                {"id": 1, "name": "Alice"},
+                {"id": 2, "name": "Bob"},
+                {"id": 3, "name": "Charlie"},
+            ],
+        )
+
+        # Test single positional parameter in WHERE clause
+        query = f"""
+        SELECT id, name
+        FROM `{self.project_id}.{address.dataset_id}.{address.table_id}`
+        WHERE id = ?
+        ORDER BY id
+        """
+
+        # Positional parameters are specified using ScalarQueryParameter with name=None
+        job_config = bigquery.QueryJobConfig(
+            query_parameters=[
+                bigquery.ScalarQueryParameter(None, "INT64", 2)
+            ]
+        )
+
+        self.run_query_test(
+            query,
+            expected_result=[
+                {"id": 2, "name": "Bob"},
+            ],
+            job_config=job_config,
+        )

